crates/l3/common/src/lib.rs

```diff
@@ -0,0 +1,113 @@
+pub mod errors;
+use color_eyre::{eyre::eyre, Result};
+use starknet_accounts::{Account, Call, ConnectedAccount, Execution, SingleOwnerAccount};
+use starknet_contract::ContractFactory;
+use starknet_core::types::contract::{CompiledClass, SierraClass};
+use starknet_core::types::{BlockId, BlockTag, FunctionCall, InvokeTransactionResult};
+use starknet_core::utils::get_selector_from_name;
+use starknet_ff::FieldElement;
+use starknet_providers::jsonrpc::{HttpTransport, JsonRpcClient};
+use starknet_providers::Provider;
+use starknet_signers::LocalWallet;
+use std::sync::Arc;
+
+pub type LocalWalletSignerMiddleware =
+    SingleOwnerAccount<JsonRpcClient<HttpTransport>, LocalWallet>;
+
+type RpcAccount<'a> = SingleOwnerAccount<&'a JsonRpcClient<HttpTransport>, LocalWallet>;
+pub type TransactionExecution<'a> = Execution<'a, RpcAccount<'a>>;
+
+pub const NO_CONSTRUCTOR_ARG: Vec<FieldElement> = Vec::new();
+
+pub const MAX_FEE: &str = "0x1000000000000";
+
+pub trait StarknetContractClient {
+    fn address(&self) -> FieldElement;
+    fn client(&self) -> LocalWalletSignerMiddleware;
+}
+
+pub async fn invoke_contract(
+    client: &Arc<&LocalWalletSignerMiddleware>,
+    address: FieldElement,
+    method: &str,
+    calldata: Vec<FieldElement>,
+) -> Result<InvokeTransactionResult> {
+    let selector = get_selector_from_name(method.into())
+        .map_err(|e| eyre!("Invalid selector for {}: {}", method, e))?;
+    let call = Call {
+        to: address,
+        selector,
+        calldata,
+    };
+    let max_fee = FieldElement::from_hex_be(MAX_FEE).unwrap();
+    client
+        .as_ref()
+        .execute(vec![call])
+        .max_fee(max_fee)
+        .send()
+        .await
+        .map_err(|e| eyre!("Failed to send transaction: {}", e))
+}
+
+pub async fn call_contract(
+    client: &LocalWalletSignerMiddleware,
+    address: FieldElement,
+    method: &str,
+    calldata: Vec<FieldElement>,
+) -> Result<Vec<FieldElement>> {
+    let entry_point_selector = get_selector_from_name(method.into())
+        .map_err(|e| eyre!("Invalid selector for {}: {}", method, e))?;
+    let function_call = FunctionCall {
+        contract_address: address,
+        entry_point_selector,
+        calldata,
+    };
+    let provider = client.provider();
+    provider
+        .call(function_call, BlockId::Tag(BlockTag::Latest))
```

Comment:
Same. It looks like you are using client only to get `provider` out of it.
Can you change the signature so that it takes the provider directly as argument?

Reply:
done